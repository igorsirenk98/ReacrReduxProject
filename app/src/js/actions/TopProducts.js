import {
    FETCH_TOP_PRODUCTS_START,
    FETCH_TOP_PRODUCTS_SUCCESS,
    FETCH_TOP_PRODUCTS_FAILURE
} from '../constants/action-types';

export function fetchTopProducts() {
    return dispatch => {
        dispatch(fetchTopProductsStart());
        return fetch('http://localhost:2000/products/top-products')
            .then(res => res.json())
            .then(res => {
                dispatch(fetchTopProductsSuccess(res));
                return res;
            })
            .catch(error => dispatch(fetchTopProductsFailure(error)));
    };
}

export const fetchTopProductsStart = products => ({
    type: FETCH_TOP_PRODUCTS_START
});

export const fetchTopProductsSuccess = products => ({
    type: FETCH_TOP_PRODUCTS_SUCCESS,
    payload: {
        products
    }
});

export const fetchTopProductsFailure = error => ({
    type: FETCH_TOP_PRODUCTS_FAILURE,
    payload: {
        error
    }
});