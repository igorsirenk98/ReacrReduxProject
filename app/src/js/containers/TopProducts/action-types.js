export const FETCH_TOP_PRODUCTS_START = 'FETCH_TOP_PRODUCTS_START';
export const FETCH_TOP_PRODUCTS_RESULT = 'FETCH_TOP_PRODUCTS_RESULT';