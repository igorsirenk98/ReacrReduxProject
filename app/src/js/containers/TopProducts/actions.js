import {
	FETCH_TOP_PRODUCTS_START,
	FETCH_TOP_PRODUCTS_RESULT
} from './action-types';

import callApi from '../../utils/callApi';

export const fetchTopProductsStart = () => ({
	type: FETCH_TOP_PRODUCTS_START,
});

export const fetchTopProductsResult = (payload, error) => ({
	type: FETCH_TOP_PRODUCTS_RESULT,
	payload,
	error
});

export function fetchTopProducts() {
	const url = `products/top-products`;

	return (dispatch) => {
		return dispatch(callApi([fetchTopProductsStart, fetchTopProductsResult], url));
	}
}
