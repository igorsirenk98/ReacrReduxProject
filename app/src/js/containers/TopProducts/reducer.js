import {
	FETCH_TOP_PRODUCTS_START,
	FETCH_TOP_PRODUCTS_RESULT
} from './action-types';

const initialState = {
	payload: {
		products: []
	},
	error: false
};

const topProducts = (state = initialState, action) => {
	switch (action.type) {
		case FETCH_TOP_PRODUCTS_START: {
			return {
				...state,
				error: false
			};
		}
		case FETCH_TOP_PRODUCTS_RESULT: {
			return {
				...state,
				payload: action.payload,
				error: action.error
			};
		}
		default:
			return state;
	}
};

export default topProducts;
