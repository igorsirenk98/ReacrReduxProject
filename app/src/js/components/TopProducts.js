import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';
import Typography from '@material-ui/core/Typography';
import Container from '@material-ui/core/Container';

import { fetchTopProducts } from '../actions/TopProducts';
import { List } from './basicComponents/List';
import Loading from './basicComponents/Loading';

const mapStateToProps = state => ({
    products: state.topProducts.products,
    loading: state.topProducts.loading,
    error: state.topProducts.error
});

class TopProducts extends Component {
    componentDidMount() {
        this.props.dispatch(fetchTopProducts());
    }

    render() {
        const { products, loading, error } = this.props;

        if (error) {
            return <div className="error"><p>Error! {error.message}</p></div>;
        } else if (loading) {
            return (
                <>
                    <CssBaseline />
                    <Container className="containerStyles" fixed>
                        <div className="loadingContainer">
                            <Loading />
                        </div>
                    </Container>
                </>
            );
        }

        return (
            <>
                {!!products.length && 
                    <List products={products}/>
                }
            </>
        ); 
    }
}

export default connect(mapStateToProps)(TopProducts);