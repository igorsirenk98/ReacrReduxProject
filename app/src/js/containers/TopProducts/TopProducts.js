import React, { Component } from 'react';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import CssBaseline from '@material-ui/core/CssBaseline';
import Typography from '@material-ui/core/Typography';
import Container from '@material-ui/core/Container';

import List from '../../components/controls/List';
import Loading from '../controls/Loading/Loading';
import ErrorMessage from '../../components/controls/ErrorMessage';

import { fetchTopProducts } from './actions';

const mapStateToProps = state => ({
    payload: state.topProducts.payload,
    isLoading: state.loadingAnimation.payload.isLoading,
    error: state.topProducts.error
});

const mapDispatchToProps = dispatch => {
    return bindActionCreators({
        fetchTopProducts
    }, dispatch)
};

export class TopProducts extends Component {
    componentDidMount() {
        this.props.fetchTopProducts();
    }

    render() {
        const { payload, isLoading, error } = this.props;
        const { products } = payload;

        if (error) {
            return <ErrorMessage error={payload} />;
        } else if (isLoading) {
            return (
                <>
                    <CssBaseline />
                    <Container className="container" fixed>
                        <div className="loading-container">
                            <Loading />
                        </div>
                    </Container>
                </>
            );
        }

        return (
            <>
                <>
                    <Typography className="notification" gutterBottom variant="h3" component="h2">
                        Top products
                    </Typography>
                    <List products={products} />
                </>
            </>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(TopProducts);