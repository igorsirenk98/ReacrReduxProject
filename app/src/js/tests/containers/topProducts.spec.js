import React from 'react';
import { shallow } from 'enzyme';
import { TopProducts } from '../../containers/TopProducts/TopProducts';

describe('TopProducts container', () => {
    const props = {
        payload: {
            products: [{name: 'bike1'}, {name: 'bike2'}]
        },
        isLoading: false,
        error: false,
        fetchTopProducts: () => { },
    }

    describe('TopProducts initial', () => {
        const topProducts = shallow(<TopProducts {...props} />);

        it('renders correctly', () => {
            expect(topProducts.debug()).toMatchSnapshot();
        })
    })

    describe('TopProducts isLoading', () => {
        const nextProps = {
            ...props,
            payload: {
                products: [{name: 'bike1'}, {name: 'bike2'}],
            },
            isLoading: true,
        }

        const topProducts = shallow(<TopProducts {...nextProps} />);

        it('renders correctly', () => {
            expect(topProducts.debug()).toMatchSnapshot();
        })
    })

    describe('TopProducts error', () => {
        const nextProps = {
            ...props,
            payload: {
                error: {
                    message: 'error',
                }
            },
            error: true,
        }

        const topProducts = shallow(<TopProducts {...nextProps} />);

        it('renders correctly', () => {
            expect(topProducts.debug()).toMatchSnapshot();
        })
    })

    describe('TopProducts with data', () => {
        const nextProps = {
            ...props,
            payload: {
                products: [{name: 'bike1'}, {name: 'bike2'}],
            }
        }

        const topProducts = shallow(<TopProducts {...nextProps} />);

        it('renders correctly', () => {
            expect(topProducts.debug()).toMatchSnapshot();
        })
    })
})

