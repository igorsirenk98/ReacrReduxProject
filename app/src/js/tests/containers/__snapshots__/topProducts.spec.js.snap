// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopProducts container TopProducts error renders correctly 1`] = `"<ErrorMessage error={{...}} />"`;

exports[`TopProducts container TopProducts initial renders correctly 1`] = `
"<Fragment>
  <WithStyles(ForwardRef(Typography)) className=\\"notification\\" gutterBottom={true} variant=\\"h3\\" component=\\"h2\\">
    Top products
  </WithStyles(ForwardRef(Typography))>
  <List products={{...}} />
</Fragment>"
`;

exports[`TopProducts container TopProducts isLoading renders correctly 1`] = `
"<Fragment>
  <CssBaseline />
  <WithStyles(ForwardRef(Container)) className=\\"container\\" fixed={true}>
    <div className=\\"loading-container\\">
      <Connect(Loading) />
    </div>
  </WithStyles(ForwardRef(Container))>
</Fragment>"
`;

exports[`TopProducts container TopProducts with data renders correctly 1`] = `
"<Fragment>
  <WithStyles(ForwardRef(Typography)) className=\\"notification\\" gutterBottom={true} variant=\\"h3\\" component=\\"h2\\">
    Top products
  </WithStyles(ForwardRef(Typography))>
  <List products={{...}} />
</Fragment>"
`;
